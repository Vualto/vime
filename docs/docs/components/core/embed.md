---
title: vm-embed
sidebar_label: Embed
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

Embeds an external media player and enables interacting with it via `postMessage`. This is generally
used internally by other providers, but you could use it if your requirements are simple. You'll
also get the benefits of preconnections and lazy-loading. Refer to [existing providers](#used-by) to
see what params you can pass in, how to send commands to the player, and how to listen to events.

<!-- Auto Generated Below -->

## Usage

<Tabs
groupId="framework"
defaultValue="html"
values={[
{ label: 'HTML', value: 'html' },
{ label: 'React', value: 'react' },
{ label: 'Vue', value: 'vue' },
{ label: 'Svelte', value: 'svelte' },
{ label: 'Stencil', value: 'stencil' },
{ label: 'Angular', value: 'angular' }
]}>

<TabItem value="html">

```html
<vm-embed
  embed-src="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
  params="autoplay=1&muted=1&controls=0"
  media-title="Agent 327: Operation Barbershop"
  origin="https://www.youtube-nocookie.com"
/>

<script>
  const embed = document.querySelector('vm-embed');

  embed.addEventListener('vmEmbedMessage', (e) => {
    const message = e.detail;
    // ...
  });
</script>
```

</TabItem>


<TabItem value="react">

```tsx {2,11-17}
import React from 'react';
import { Embed } from '@vime/react';

function Example() {
  const onMessage = (event: CustomEvent<any>) => {
    const message = event.detail;
    // ...
  };

  return (
    <Embed
      embedSrc="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
      params={{ autoplay: 1, muted: 1, controls: 0 }}
      mediaTitle="Agent 327: Operation Barbershop"
      origin="https://www.youtube-nocookie.com"
      onVmEmbedMessage={onMessage}
    />
  );
}
```

</TabItem>


<TabItem value="vue">

```html {2-8,12,16} title="example.vue"
<template>
  <embed
    embedSrc="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
    mediaTitle="Agent 327: Operation Barbershop"
    origin="https://www.youtube-nocookie.com"
    :params="params"
    @vmEmbedMessage="onMessage"
  />
</template>

<script>
  import { Embed } from '@vime/vue';

  export default {
    components: {
      Embed,
    },

    data: {
      params: {
        autoplay: 1,
        muted: 1,
        controls: 0,
      },
    },

    methods: {
      onMessage(message: any) {
        // ...
      },
    },
  };
</script>
```

</TabItem>


<TabItem value="svelte">

```tsx
<Embed
  embedSrc="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
  origin="https://www.youtube-nocookie.com"
  mediaTitle="Agent 327: Operation Barbershop"
  params={params}
  on:vmEmbedMessage={onMessage}
/>
```

```html {2}
<script lang="ts">
  import { Embed } from '@vime/svelte';

  const params = {
    autoplay: 1,
    muted: 1,
    controls: 0,
  };

  const onMessage = (event: CustomEvent<any>) => {
    const message = event.detail;
    // ...
  };
</script>
```

</TabItem>


<TabItem value="stencil">

```tsx
class Example {
  private onMessage(event: CustomEvent<any>) {
    const message = event.detail;
    // ...
  }

  render() {
    return (
      <vm-embed
        embedSrc="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
        params={{ autoplay: 1, muted: 1, controls: 0 }}
        mediaTitle="Agent 327: Operation Barbershop"
        origin="https://www.youtube-nocookie.com"
        onVmEmbedMessage={this.onMessage.bind(this)}
      />
    );
  }
}
```

</TabItem>


<TabItem value="angular">

```html title="example.html"
<vm-embed
  embed-src="https://www.youtube-nocookie.com/embed/DyTCOwB0DVw"
  media-title="Agent 327: Operation Barbershop"
  origin="https://www.youtube-nocookie.com"
  [params]="params"
  (vmEmbedMessage)="onMessage($event)"
/>
```

```ts title="example.ts"
class Example {
  params = {
    autoplay: 1,
    muted: 1,
    controls: 0,
  };

  onMessage(event: CustomEvent<any>) {
    const message = event.detail;
    // ...
  }
}
```

</TabItem>
    
</Tabs>


## Properties

| Property         | Attribute     | Description                                                                                                                                  | Type                                                              | Default     |
| ---------------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------- | ----------- |
| `decoder`        | --            | A function which accepts the raw message received from the embedded media player via `postMessage` and converts it into a POJO.              | `((data: string) => Record<string, any> ∣ undefined) ∣ undefined` | `undefined` |
| `embedSrc`       | `embed-src`   | A URL that will load the external player and media (Eg: https://www.youtube.com/embed/DyTCOwB0DVw).                                          | `string`                                                          | `''`        |
| `mediaTitle`     | `media-title` | The title of the current media so it can be set on the inner `iframe` for screen readers.                                                    | `string`                                                          | `''`        |
| `origin`         | `origin`      | Where the src request had originated from without any path information.                                                                      | `string ∣ undefined`                                              | `undefined` |
| `params`         | `params`      | The parameters to pass to the embedded player which are appended to the `embedSrc` prop. These can be passed in as a query string or object. | `string ∣ { [x: string]: any; }`                                  | `''`        |
| `preconnections` | --            | A collection of URLs to that the browser should immediately start establishing a connection with.                                            | `string[]`                                                        | `[]`        |

## Events

| Event              | Description                                                                                                                                        | Type                  |
| ------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------- |
| `vmEmbedLoaded`    | Emitted when the embedded player and any new media has loaded.                                                                                     | `CustomEvent<void>`   |
| `vmEmbedMessage`   | Emitted when a new message is received from the embedded player via `postMessage`.                                                                 | `CustomEvent<any>`    |
| `vmEmbedSrcChange` | Emitted when the `embedSrc` or `params` props change. The payload contains the `params` serialized into a query string and appended to `embedSrc`. | `CustomEvent<string>` |

## Methods

### `postMessage(message: any, target?: string | undefined) => Promise<void>`

Posts a message to the embedded media player.

#### Returns

Type: `Promise<void>`

## Dependencies

### Used by

- [vm-dailymotion](../providers/dailymotion.md)
- [vm-vimeo](../providers/vimeo.md)
- [vm-youtube](../providers/youtube.md)

### Graph

```mermaid
graph TD;
  vm-dailymotion --> vm-embed
  vm-vimeo --> vm-embed
  vm-youtube --> vm-embed
  style vm-embed fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
