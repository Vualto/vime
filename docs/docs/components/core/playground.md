---
title: vm-playground
sidebar_label: Playground
---

A simple playground for testing and playing with Vime and its various providers.

<!-- Auto Generated Below -->

## Properties

| Property       | Attribute         | Description                                                         | Type                                                                                                                                            | Default                        |
| -------------- | ----------------- | ------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ |
| `poster`       | `poster`          | The current poster to load.                                         | `string`                                                                                                                                        | `${BASE_MEDIA_URL}/poster.png` |
| `provider`     | `provider`        | The current media provider.                                         | `Provider.Audio ∣ Provider.Dailymotion ∣ Provider.Dash ∣ Provider.FakeTube ∣ Provider.HLS ∣ Provider.Video ∣ Provider.Vimeo ∣ Provider.YouTube` | `Provider.Audio`               |
| `showControls` | `show-controls`   | Whether to show the native controls or not.                         | `boolean`                                                                                                                                       | `true`                         |
| `showCustomUI` | `show-custom-u-i` | Whether to show the custom Vime UI or not.                          | `boolean`                                                                                                                                       | `false`                        |
| `src`          | `src`             | The current `src` to load into the provider.                        | `string ∣ undefined`                                                                                                                            | `undefined`                    |
| `theme`        | `theme`           | The current custom UI theme, won't work if custom UI is turned off. | `"dark" ∣ "light"`                                                                                                                              | `'dark'`                       |

## Dependencies

### Depends on

- [vm-audio](../providers/audio.md)
- [vm-video](../providers/video.md)
- [vm-hls](../providers/hls.md)
- [vm-dash](../providers/dash.md)
- [vm-youtube](../providers/youtube.md)
- [vm-vimeo](../providers/vimeo.md)
- [vm-dailymotion](../providers/dailymotion.md)
- [vm-player](player.md)
- [vm-default-ui](../ui/default-ui.md)

### Graph

```mermaid
graph TD;
  vm-playground --> vm-audio
  vm-playground --> vm-video
  vm-playground --> vm-hls
  vm-playground --> vm-dash
  vm-playground --> vm-youtube
  vm-playground --> vm-vimeo
  vm-playground --> vm-dailymotion
  vm-playground --> vm-player
  vm-playground --> vm-default-ui
  vm-audio --> vm-file
  vm-video --> vm-file
  vm-hls --> vm-video
  vm-dash --> vm-video
  vm-youtube --> vm-embed
  vm-vimeo --> vm-embed
  vm-dailymotion --> vm-embed
  vm-default-ui --> vm-ui
  vm-default-ui --> vm-icon-library
  vm-default-ui --> vm-skeleton
  vm-default-ui --> vm-click-to-play
  vm-default-ui --> vm-dbl-click-fullscreen
  vm-default-ui --> vm-captions
  vm-default-ui --> vm-poster
  vm-default-ui --> vm-spinner
  vm-default-ui --> vm-default-controls
  vm-default-ui --> vm-default-settings
  vm-default-controls --> vm-controls
  vm-default-controls --> vm-playback-control
  vm-default-controls --> vm-volume-control
  vm-default-controls --> vm-current-time
  vm-default-controls --> vm-control-spacer
  vm-default-controls --> vm-scrubber-control
  vm-default-controls --> vm-live-indicator
  vm-default-controls --> vm-end-time
  vm-default-controls --> vm-settings-control
  vm-default-controls --> vm-control-group
  vm-default-controls --> vm-fullscreen-control
  vm-default-controls --> vm-scrim
  vm-default-controls --> vm-caption-control
  vm-default-controls --> vm-time-progress
  vm-default-controls --> vm-pip-control
  vm-playback-control --> vm-control
  vm-playback-control --> vm-icon
  vm-playback-control --> vm-tooltip
  vm-volume-control --> vm-mute-control
  vm-volume-control --> vm-slider
  vm-mute-control --> vm-control
  vm-mute-control --> vm-icon
  vm-mute-control --> vm-tooltip
  vm-current-time --> vm-time
  vm-scrubber-control --> vm-slider
  vm-scrubber-control --> vm-tooltip
  vm-end-time --> vm-time
  vm-settings-control --> vm-control
  vm-settings-control --> vm-icon
  vm-settings-control --> vm-tooltip
  vm-fullscreen-control --> vm-control
  vm-fullscreen-control --> vm-icon
  vm-fullscreen-control --> vm-tooltip
  vm-time-progress --> vm-current-time
  vm-time-progress --> vm-end-time
  vm-pip-control --> vm-control
  vm-pip-control --> vm-icon
  vm-pip-control --> vm-tooltip
  vm-default-settings --> vm-menu-item
  vm-default-settings --> vm-menu-radio
  vm-default-settings --> vm-submenu
  vm-default-settings --> vm-menu-radio-group
  vm-default-settings --> vm-settings
  vm-menu-item --> vm-icon
  vm-menu-radio --> vm-menu-item
  vm-submenu --> vm-menu-item
  vm-submenu --> vm-menu
  vm-settings --> vm-menu
  style vm-playground fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
