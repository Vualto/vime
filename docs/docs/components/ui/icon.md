---
title: vm-icon
sidebar_label: Icon
---

> The implementation was inspired by [Shoelace](https://github.com/shoelace-style/shoelace), thanks Cory!

Loads and renders an SVG icon. The SVG can be loaded locally or through a CORS enabled endpoint. If
you have registered an icon library, you can resolve the icon by name.

<!-- Auto Generated Below -->

## Usage

<Tabs
groupId="framework"
defaultValue="html"
values={[
{ label: 'HTML', value: 'html' },
{ label: 'React', value: 'react' },
{ label: 'Vue', value: 'vue' },
{ label: 'Svelte', value: 'svelte' },
{ label: 'Stencil', value: 'stencil' },
{ label: 'Angular', value: 'angular' }
]}>

<TabItem value="html">

```html
<!-- Src. -->
<vm-icon src="/icons/my-icon.svg" label="An icon"></vm-icon>

<!-- Icon library. -->
<vm-icon name="pause" library="material" label="Pause"></vm-icon>
```

</TabItem>


<TabItem value="react">

```tsx {2,8,11}
import React from 'react';
import { Icon } from '@vime/react';

function Example() {
  return (
    <div>
      {/* Src. */}
      <Icon src="/icons/my-icon.svg" label="An icon" />

      {/* Icon library. */}
      <Icon name="play" library="material" label="Play" />
    </div>
  );
}
```

</TabItem>


<TabItem value="vue">

```html {4,7,12,16} title="example.vue"
<template>
  <div>
    <!-- Src. -->
    <Icon src="/icons/my-icon.svg" label="An icon" />

    <!-- Icon library. -->
    <Icon name="pause" library="material" label="Pause" />
  </div>
</template>

<script>
  import { Icon } from '@vime/vue';

  export default {
    components: {
      Icon,
    },
  };
</script>
```

</TabItem>


<TabItem value="svelte">

```html {2,5,8} title="example.svelte"
<!-- Src. -->
<Icon src="/icons/my-icon.svg" label="An icon" />

<!-- Icon library. -->
<Icon name="pause" library="material" label="Pause" />

<script lang="ts">
  import { Icon } from '@vime/svelte';
<script>
```

</TabItem>


<TabItem value="stencil">

```tsx {6,9}
class Example {
  render() {
    return (
      <div>
        {/* Src. */}
        <vm-icon src="/icons/my-icon.svg" label="An icon" />

        {/* Icon library. */}
        <vm-icon name="play" library="material" label="Play" />
      </div>
    );
  }
}
```

</TabItem>


<TabItem value="angular">

```html title="example.html"
<!-- Src. -->
<vm-icon src="/icons/my-icon.svg" label="An icon"></vm-icon>

<!-- Icon library. -->
<vm-icon name="play" library="material" label="Play"></vm-icon>
```

</TabItem>
    
</Tabs>


## Properties

| Property  | Attribute | Description                                                                                                   | Type                 | Default     |
| --------- | --------- | ------------------------------------------------------------------------------------------------------------- | -------------------- | ----------- |
| `label`   | `label`   | An alternative description to use for accessibility. If omitted, the name or src will be used to generate it. | `string ∣ undefined` | `undefined` |
| `library` | `library` | The name of a registered icon library.                                                                        | `string ∣ undefined` | `undefined` |
| `name`    | `name`    | The name of the icon to draw.                                                                                 | `string ∣ undefined` | `undefined` |
| `src`     | `src`     | The absolute URL of an SVG file to load.                                                                      | `string ∣ undefined` | `undefined` |

## Events

| Event     | Description                           | Type                                            |
| --------- | ------------------------------------- | ----------------------------------------------- |
| `vmError` | Emitted when the icon failed to load. | `CustomEvent<{ status?: number ∣ undefined; }>` |
| `vmLoad`  | Emitted when the icon has loaded.     | `CustomEvent<void>`                             |

## Dependencies

### Used by

- [vm-fullscreen-control](controls/fullscreen-control.md)
- [vm-menu-item](settings/menu-item.md)
- [vm-mute-control](controls/mute-control.md)
- [vm-pip-control](controls/pip-control.md)
- [vm-playback-control](controls/playback-control.md)
- [vm-settings-control](controls/settings-control.md)

### Graph

```mermaid
graph TD;
  vm-fullscreen-control --> vm-icon
  vm-menu-item --> vm-icon
  vm-mute-control --> vm-icon
  vm-pip-control --> vm-icon
  vm-playback-control --> vm-icon
  vm-settings-control --> vm-icon
  style vm-icon fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
