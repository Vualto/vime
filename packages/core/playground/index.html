<!DOCTYPE html>
<html dir="ltr">

<head>
  <meta charset="UTF-8">
  <title>Playground | Vime</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no">

  <link rel="stylesheet" href="/src/themes/default.css">
  <link rel="stylesheet" href="/src/themes/light.css">
  <link rel="icon" type="image/png" href="/playground/favicon.ico">
  <script type="module" src="/dist/vime/vime.esm.js"></script>

  <style>
    html, body {
      width: 100%;
      height: 100%;
      margin: 0;
      box-sizing: border-box;
    }

    body {
      padding: 16px;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
    }
  </style>
</head>
<body>
  <vm-playground></vm-playground>
</body>
</html>
