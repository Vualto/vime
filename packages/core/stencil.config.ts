import { Config } from '@stencil/core';
import { svelteOutputTarget } from './targets/svelte';
import { reactOutputTarget } from './targets/react';
import { angularOutputTarget } from './targets/angular';
import { vueOutputTarget } from './targets/vue';
import { vueNextOutputTarget } from './targets/vue-next';

export const config: Config = {
  namespace: 'Vime',
  taskQueue: 'async',
  outputTargets: [
    reactOutputTarget(),
    angularOutputTarget(),
    vueOutputTarget(),
    vueNextOutputTarget(),
    svelteOutputTarget(),
    {
      type: 'dist',
      esmLoaderPath: '../loader',
    },
    {
      type: 'dist-custom-elements-bundle',
    },
  ],
  testing: {
    globals: {
      'ts-jest': {
        diagnostics: false,
      },
    },
    setupFilesAfterEnv: [
      './src/globals/jest.js',
    ],
    coverageDirectory: 'jest-coverage',
    // @ts-ignore
    watchPlugins: [
      'jest-watch-typeahead/filename',
      'jest-watch-typeahead/testname',
    ],
    testPathIgnorePatterns: [
      '/node_modules/',
      '/cypress/',
    ],
  },
  preamble: '(C) Vime https://vimejs.com - MIT License',
};
