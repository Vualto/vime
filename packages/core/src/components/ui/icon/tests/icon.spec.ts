import { newSpecPage } from '@stencil/core/testing';
import { Icon } from '../icon';

it('should be structurally sound', async () => {
  const page = await newSpecPage({
    components: [Icon],
    html: '<vm-icon name="icon"></vm-icon>',
  });

  expect(page.root).toMatchSnapshot();
});
