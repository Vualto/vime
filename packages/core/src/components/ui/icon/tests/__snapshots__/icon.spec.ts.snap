// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept svg markup 1`] = `
<vm-icon>
  <!---->
  <svg focusable="false" role="presentation" xmlns="http://www.w3.org/2000/svg">
    <rect height="100" width="300"></rect>
  </svg>
</vm-icon>
`;

exports[`should be structurally sound 1`] = `
<vm-icon href="#my-icon">
  <!---->
  <svg focusable="false" role="presentation" xmlns="http://www.w3.org/2000/svg">
    <use href="#my-icon"></use>
  </svg>
</vm-icon>
`;
